ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/participant/state/kvutils/committer/transaction/TransactionCommitterSpec.scala

```diff
@@ -595,58 +595,125 @@ class TransactionCommitterSpec extends AnyWordSpec with Matchers with MockitoSug
+
+      val cases =
+        Seq((false, Some(s"#$freshContractId")), (true, Some(s"#$freshContractId")), (true, None))
+          .flatMap { case (found, contractIdAtCommitter) =>
+            Seq(false, true).map(inRollback =>
+              newLookupByKeySubmittedTransaction(found, inRollback) -> contractIdAtCommitter
+            )
+          }
 
-      val cases = Table(
+      val casesTbl = Table(
         "transaction" -> "contractIdAtCommitter",
-        lookupByKeyNotFound -> Some(s"#$freshContractId"),
-        lookupByKeyFound1 -> Some(s"#$freshContractId"),
-        lookupByKeyFound2 -> None,
+        cases: _*
       )
 
-      val key = {
-        val aCreateNode = createNode("#dummy")
-        Conversions.encodeContractKey(aCreateNode.templateId, aCreateNode.key.get.key)
-      }
-      forAll(cases) { (transaction: SubmittedTransaction, contractIdAtCommitter: Option[String]) =>
-        val context = commitContextWithContractStateKeys(
-          key -> contractIdAtCommitter
-        )
-        val result =
-          ContractKeysValidation
-            .validateKeys(transactionCommitter)
-            .apply(
-              context,
-              DamlTransactionEntrySummary(createTransactionEntry(List("Alice"), transaction)),
-            )
-        result shouldBe a[StepStop]
-        val rejectionReason =
-          getTransactionRejectionReason(result).getInconsistent.getDetails
-        rejectionReason should startWith("InconsistentKeys")
+      forAll(casesTbl) {
+        (transaction: SubmittedTransaction, contractIdAtCommitter: Option[String]) =>
+          val context = commitContextWithContractStateKeys(
+            mykey -> contractIdAtCommitter
+          )
+          val result = validate(context, transaction)
+          result shouldBe a[StepStop]
+          val rejectionReason =
+            getTransactionRejectionReason(result).getInconsistent.getDetails
+          rejectionReason should startWith("InconsistentKeys")
       }
     }
 
+    "return DuplicateKeys when two local contracts conflict" in {
+      val builder = TransactionBuilder()
+      builder.add(createNode(s"#$freshContractId"))
+      builder.add(createNode(s"#$freshContractId"))
+      val transaction = builder.buildSubmitted()
+      val context = commitContextWithContractStateKeys(mykey -> None)
+      val result = validate(context, transaction)
+      result shouldBe a[StepStop]
+      val rejectionReason =
+        getTransactionRejectionReason(result).getInconsistent.getDetails
+      rejectionReason should startWith("DuplicateKeys")
+    }
+
+    "return DuplicateKeys when a create in a rollback conflicts with a global key" in {
+      val builder = TransactionBuilder()
+      val rollback = builder.add(builder.rollback())
+      builder.add(createNode(s"#$freshContractId"), rollback)
+      val transaction = builder.buildSubmitted()
+      val context = commitContextWithContractStateKeys(mykey -> Some(s"#freshContractId"))
```

Comment:
Not that it really matters, but for consistency/readability:

```suggestion
      val context = commitContextWithContractStateKeys(mykey -> Some(s"#$freshContractId"))
```